File: typescript/src/structured.ts
import type { JsonValue } from "./model/types.js";
import { copyJsonValue } from "./model/validation.js";
import { AgentError } from "./shared/errors.js";
import { findSchemaViolation, inspectSchemaNode, type JsonSchema } from "./shared/schema.js";

export const OUTPUT_SCHEMA_PROFILE = "purra.output-schema/v1" as const;
export const JSON_IDENTITY_PROFILE = "purra.json-identity/v1" as const;

export interface StructuredOutputLimits {
  readonly schemaBytes: number;
  readonly outputBytes: number;
  readonly schemaDepth: number;
  readonly outputDepth: number;
  readonly schemaNodes: number;
  readonly outputNodes: number;
  readonly validationSteps: number;
}

const DEFAULT_LIMITS: StructuredOutputLimits = Object.freeze({
  schemaBytes: 65_536, outputBytes: 1_048_576,
  schemaDepth: 32, outputDepth: 64, schemaNodes: 4_096,
  outputNodes: 65_536, validationSteps: 100_000,
});

export interface StructuredOutputOptions {
  readonly schemaId: string;
  readonly schemaVersion: string;
  readonly schema: JsonSchema;
  readonly mode?: "local" | "native_required";
  readonly schemaProfile?: typeof OUTPUT_SCHEMA_PROFILE;
  readonly limits?: Partial<StructuredOutputLimits>;
}

export class StructuredOutputError extends AgentError {
  public readonly details: Readonly<{ path: string; keyword: string; reason: string }>;
  public constructor(code: string, keyword: string, path = "$") {
    super(code, "Structured output contract validation failed");
    this.name = "StructuredOutputError";
    this.details = Object.freeze({ path: encoder.encode(path).length <= 512 ? path : "$", keyword, reason: "constraint_failed" });
  }
}

const encoder = new TextEncoder();
function validString(value: string): void {
  // TextEncoder replaces lone surrogates; reject them before measuring/hashing.
  for (const char of value) {
    const point = char.codePointAt(0)!;
    if (point >= 0xd800 && point <= 0xdfff) throw new Error("invalid Unicode scalar");
  }
}

function snapshot(value: unknown, depth: number, nodes: number, byteLimit: number): JsonValue {
  let remaining = nodes;
  let bytes = 0;
  const active = new Set<object>();
  function visit(item: unknown, level: number): JsonValue {
    if (--remaining < 0 || level > depth) throw new Error("JSON resource limit exceeded");
    let result: JsonValue;
    if (typeof item === "string") {
      if (item.length > byteLimit) throw new Error("JSON byte limit exceeded");
      validString(item);
      bytes += encoder.encode(JSON.stringify(item)).length;
      result = item;
    } else if (item === null || typeof item === "boolean") {
      bytes += 5;
      result = item;
    } else if (typeof item === "number") {
      if (!Number.isFinite(item) || (Number.isInteger(item) && !Number.isSafeInteger(item))) {
        throw new Error("number outside interoperable range");
      }
      bytes += 24;
      result = item === 0 ? 0 : item;
    } else if (typeof item === "object" && item !== null) {
      if (active.has(item)) throw new Error("cyclic JSON");
      if (!Array.isArray(item) && ![Object.prototype, null].includes(Object.getPrototypeOf(item))) {
        throw new Error("unsupported JSON object");
      }
      active.add(item);
      try {
        if (Array.isArray(item)) {
          result = Array.from(item, (child) => visit(child, level + 1));
        } else {
          result = Object.fromEntries(Object.entries(item).map(([key, child]) => {
            visit(key, level + 1);
            return [key, visit(child, level + 1)];
          }));
        }
        bytes += 2 + 2 * Object.keys(item).length;
      } finally { active.delete(item); }
    } else throw new Error("unsupported JSON value");
    if (bytes > byteLimit) throw new Error("JSON byte limit exceeded");
    return result;
  }
  const result = visit(value, 0);
  return result;
}

function hex(data: Uint8Array): string {
  return [...data].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function identity(value: JsonValue): string {
  if (value === null) return "n";
  if (typeof value === "boolean") return value ? "t" : "f";
  if (typeof value === "number") {
    const buffer = new ArrayBuffer(8);
    new DataView(buffer).setFloat64(0, value === 0 ? 0 : value, false);
    return "d" + hex(new Uint8Array(buffer));
  }
  if (typeof value === "string") {
    const data = encoder.encode(value);
    return `s${data.length}:${hex(data)}`;
  }
  if (Array.isArray(value)) return `a${value.length}:` + value.map(identity).join("");
  const mapping = value as JsonSchema;
  const keys = Object.keys(mapping).sort((a, b) => {
    const left = hex(encoder.encode(a)), right = hex(encoder.encode(b));
    return left < right ? -1 : left > right ? 1 : 0;
  });
  return `o${keys.length}:` + keys.map((key) => identity(key) + identity(mapping[key]!)).join("");
}

async function digest(value: JsonValue): Promise<string> {
  return hex(new Uint8Array(await globalThis.crypto.subtle.digest(
    "SHA-256", encoder.encode(JSON_IDENTITY_PROFILE + "\n" + identity(value)),
  )));
}

function parseDocument(text: string, limits: StructuredOutputLimits): JsonValue {
  if (typeof text !== "string" || text.length > limits.outputBytes
    || encoder.encode(text).length > limits.outputBytes) throw new Error("invalid document");
  let offset = 0;
  let nodes = 0;
  function white(): void { while (/[\x20\t\r\n]/.test(text[offset] ?? "!") && offset < text.length) offset++; }
  function string(): string {
    const start = offset++;
    while (offset < text.length) {
      const char = text[offset++];
      if (char === "\\") offset++;
      else if (char === '"') {
        const result = JSON.parse(text.slice(start, offset)) as string;
        validString(result);
        return result;
      }
    }
    throw new Error("unterminated string");
  }
  function value(level: number): JsonValue {
    if (++nodes > limits.outputNodes || level > limits.outputDepth) throw new Error("resource limit");
    white();
    const char = text[offset];
    if (char === '"') return string();
    if (char === "{" || char === "[") {
      offset++;
      const object = char === "{";
      const close = object ? "}" : "]";
      const entries: [string, JsonValue][] = [];
      const array: JsonValue[] = [];
      const keys = new Set<string>();
      white();
      if (text[offset] === close) { offset++; return object ? {} : []; }
      while (true) {
        white();
        if (object) {
          if (text[offset] !== '"' || ++nodes > limits.outputNodes) throw new Error("invalid key");
          const key = string();
          if (keys.has(key)) throw new Error("duplicate key");
          keys.add(key);
          white();
          if (text[offset++] !== ":") throw new Error("missing colon");
          entries.push([key, value(level + 1)]);
        } else array.push(value(level + 1));
        white();
        const delimiter = text[offset++];
        if (delimiter === close) break;
        if (delimiter !== ",") throw new Error("missing delimiter");
      }
      return object ? Object.fromEntries(entries) : array;
    }
    const token = /^(?:true|false|null|-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?)/.exec(text.slice(offset))?.[0];
    if (token === undefined) throw new Error("invalid token");
    offset += token.length;
    return JSON.parse(token) as JsonValue;
  }
  const result = value(0);
  white();
  if (offset !== text.length) throw new Error("trailing content");
  return snapshot(result, limits.outputDepth, limits.outputNodes, limits.outputBytes);
}

/** Immutable output schema identity. Construction performs no external calls. */
export class StructuredOutputContract {
  public readonly schemaId: string;
  public readonly schemaVersion: string;
  public readonly schemaProfile = OUTPUT_SCHEMA_PROFILE;
  public readonly schema: JsonSchema;
  public readonly mode: "local" | "native_required";
  public readonly limits: StructuredOutputLimits;
  public readonly schemaDigest: string;
  public readonly contractDigest: string;

  private constructor(options: StructuredOutputOptions, schema: JsonSchema,
    limits: StructuredOutputLimits, schemaDigest: string, contractDigest: string) {
    this.schemaId = options.schemaId;
    this.schemaVersion = options.schemaVersion;
    this.schema = schema;
    this.mode = options.mode ?? "local";
    this.limits = limits;
    this.schemaDigest = schemaDigest;
    this.contractDigest = contractDigest;
    Object.freeze(this);
  }

  public static async create(options: StructuredOutputOptions): Promise<StructuredOutputContract> {
    let schema: JsonSchema;
    let limits: StructuredOutputLimits;
    // Capture all caller-controlled fields before the first await.
    const input = { ...options };
    try {
      for (const item of [input.schemaId, input.schemaVersion]) {
        if (typeof item !== "string" || item.length > 128 || !item.trim() || encoder.encode(item).length > 128) throw new Error("identity");
        validString(item);
      }
      if (input.schemaProfile !== undefined && input.schemaProfile !== OUTPUT_SCHEMA_PROFILE) throw new Error("profile");
      limits = Object.freeze({ ...DEFAULT_LIMITS, ...input.limits });
      if (Object.keys(limits).some((key) => !Object.hasOwn(DEFAULT_LIMITS, key))) throw new Error("limits");
      for (const key of Object.keys(DEFAULT_LIMITS) as (keyof StructuredOutputLimits)[]) {
        if (!Number.isSafeInteger(limits[key]) || limits[key] < 1 || limits[key] > DEFAULT_LIMITS[key]) throw new Error("limits");
      }
      const value = snapshot(input.schema, limits.schemaDepth, limits.schemaNodes, limits.schemaBytes);
      if (value === null || typeof value !== "object" || Array.isArray(value) || (value as JsonSchema).type !== "object") throw new Error("root");
      inspectSchemaNode(value, "$", "structured-output");
      schema = copyJsonValue(value) as JsonSchema;
    } catch {
      throw new StructuredOutputError("structured_output_schema_invalid", "schema");
    }
    if (input.mode !== undefined && input.mode !== "local" && input.mode !== "native_required") {
      throw new StructuredOutputError("structured_output_mode_unsupported", "mode");
    }
    const schemaDigest = await digest(schema);
    const contractDigest = await digest({ schemaId: input.schemaId, schemaVersion: input.schemaVersion,
      schemaProfile: OUTPUT_SCHEMA_PROFILE, schemaDigest, mode: input.mode ?? "local",
      limits: Object.keys(DEFAULT_LIMITS).map((key) => limits[key as keyof StructuredOutputLimits]),
    });
    return new StructuredOutputContract(input, schema, limits, schemaDigest, contractDigest);
  }

  public identity(): Readonly<Record<string, JsonValue>> {
    return Object.freeze({ schemaId: this.schemaId, schemaVersion: this.schemaVersion,
      schemaProfile: this.schemaProfile, schemaDigest: this.schemaDigest,
      contractDigest: this.contractDigest, mode: this.mode });
  }

  public instruction(): string {
    const header = "Return exactly one complete JSON object. Do not use tools, markdown fences, or extra text.";
    return header + (this.mode === "local" ? " JSON Schema: " + JSON.stringify(this.schema)
      : " Follow the native JSON Schema output format.");
  }

  public parse(text: string): JsonSchema {
    let value: JsonValue;
    try { value = parseDocument(text, this.limits); }
    catch { throw new StructuredOutputError("structured_output_invalid_json", "document"); }
    return this.validateValue(value);
  }

  public validateValue(input: unknown): JsonSchema {
    let value: JsonValue;
    try { value = snapshot(input, this.limits.outputDepth, this.limits.outputNodes, this.limits.outputBytes); }
    catch { throw new StructuredOutputError("structured_output_invalid_json", "value"); }
    let violation;
    try {
      violation = findSchemaViolation(value, this.schema, "$", true, { remaining: this.limits.validationSteps });
    } catch { throw new StructuredOutputError("structured_output_validation_limit_exceeded", "validation_steps"); }
    if (violation !== undefined) {
      throw new StructuredOutputError("structured_output_schema_mismatch", violation.keyword,
        violation.keyword === "additionalProperties" ? "$" : violation.path);
    }
    return copyJsonValue(value) as JsonSchema;
  }
}


/** Hash bounded JSON with the same numeric and Unicode identity as schemas. */
export async function jsonIdentityDigest(value: unknown): Promise<string> {
  return digest(snapshot(value, DEFAULT_LIMITS.outputDepth, DEFAULT_LIMITS.outputNodes, DEFAULT_LIMITS.outputBytes));
}

File: src/purra/structured.py
"""Versioned, bounded object-output contracts; no model calls or coercion."""

from __future__ import annotations

import hashlib
import json
import math
import struct
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any, Literal

from purra.errors import CodedAgentCoreError
from purra.json_values import freeze_json_mapping
from purra.schema import schema_contract_violations, schema_violation

OUTPUT_SCHEMA_PROFILE = "purra.output-schema/v1"
JSON_IDENTITY_PROFILE = "purra.json-identity/v1"
_SAFE_INTEGER = 2**53 - 1


class StructuredOutputError(CodedAgentCoreError):
    """Failure with candidate-free diagnostics; never contains rejected output."""


@dataclass(frozen=True, slots=True)
class StructuredOutputLimits:
    schema_bytes: int = 65_536
    output_bytes: int = 1_048_576
    schema_depth: int = 32
    output_depth: int = 64
    schema_nodes: int = 4_096
    output_nodes: int = 65_536
    validation_steps: int = 100_000

    def __post_init__(self) -> None:
        maxima = (65_536, 1_048_576, 32, 64, 4_096, 65_536, 100_000)
        for name, maximum in zip(self.__dataclass_fields__, maxima):
            value = getattr(self, name)
            if type(value) is not int or not 1 <= value <= maximum:
                raise ValueError(f"{name} must be a positive integer at most {maximum}")


def _failure(code: str, keyword: str, path: str = "$") -> StructuredOutputError:
    return StructuredOutputError(
        "Structured output contract validation failed",
        code=code,
        details={"path": path if len(path.encode("utf-8")) <= 512 else "$",
                 "keyword": keyword, "reason": "constraint_failed"},
    )


def _snapshot(value: Any, *, depth: int, nodes: int, byte_limit: int) -> Any:
    remaining = nodes
    # Bound detached JSON storage before creating containers; numbers reserve
    # 24 bytes and separators reserve two bytes per child in both languages.
    byte_count = 0
    active: set[int] = set()

    def visit(item: Any, level: int) -> Any:
        nonlocal remaining, byte_count
        remaining -= 1
        if remaining < 0 or level > depth:
            raise ValueError("JSON resource limit exceeded")
        if isinstance(item, str):
            if len(item) > byte_limit:
                raise ValueError("JSON byte limit exceeded")
            byte_count += len(json.dumps(item, ensure_ascii=False).encode("utf-8"))
            result = item
        elif item is None or isinstance(item, bool):
            byte_count += 5
            result = item
        elif isinstance(item, (int, float)):
            number = float(item)
            if not math.isfinite(number) or (number.is_integer() and abs(number) > _SAFE_INTEGER):
                raise ValueError("number outside interoperable range")
            result = int(number) if number.is_integer() else number
            byte_count += 24  # Conservative binary64 JSON spelling bound.
        elif isinstance(item, Mapping) or (isinstance(item, Sequence) and not isinstance(item, (bytes, bytearray))):
            if id(item) in active:
                raise ValueError("cyclic JSON")
            active.add(id(item))
            try:
                if isinstance(item, Mapping):
                    result = {}
                    for key, child in item.items():
                        if not isinstance(key, str):
                            raise ValueError("non-string JSON key")
                        visit(key, level + 1)
                        result[key] = visit(child, level + 1)
                else:
                    result = [visit(child, level + 1) for child in item]
                byte_count += 2 + 2 * len(item)
            finally:
                active.remove(id(item))
        else:
            raise ValueError("unsupported JSON value")
        if byte_count > byte_limit:
            raise ValueError("JSON byte limit exceeded")
        return result

    result = visit(value, 0)
    return result


def _identity(value: Any) -> str:
    if value is None:
        return "n"
    if isinstance(value, bool):
        return "t" if value else "f"
    if isinstance(value, (float, int)):
        return "d" + struct.pack(">d", float(value) if value else 0.0).hex()
    if isinstance(value, str):
        data = value.encode("utf-8")
        return f"s{len(data)}:{data.hex()}"
    if isinstance(value, Mapping):
        return f"o{len(value)}:" + "".join(_identity(key) + _identity(value[key]) for key in sorted(value))
    return f"a{len(value)}:" + "".join(_identity(item) for item in value)


def _digest(value: Any) -> str:
    return hashlib.sha256((JSON_IDENTITY_PROFILE + "\n" + _identity(value)).encode("ascii")).hexdigest()


def _check_schema_nodes(schema: Mapping[str, Any]) -> None:
    # The legacy tool inspector treats null as omission for some keywords.
    # Output schemas require every present keyword to have its declared type.
    for key, value in schema.items():
        if value is None and key != "const":
            raise ValueError("null schema keyword")
    properties = schema.get("properties", {})
    if isinstance(properties, Mapping):
        for child in properties.values():
            if isinstance(child, Mapping):
                _check_schema_nodes(child)
    for key in ("anyOf", "oneOf"):
        branches = schema.get(key, [])
        if isinstance(branches, list):
            for child in branches:
                if isinstance(child, Mapping):
                    _check_schema_nodes(child)
    if isinstance(schema.get("items"), Mapping):
        _check_schema_nodes(schema["items"])


@dataclass(frozen=True, slots=True)
class StructuredOutputContract:
    schema_id: str
    schema_version: str
    schema: Mapping[str, Any]
    mode: Literal["local", "native_required"] = "local"
    schema_profile: str = OUTPUT_SCHEMA_PROFILE
    limits: StructuredOutputLimits = field(default_factory=StructuredOutputLimits)
    schema_digest: str = field(init=False)
    contract_digest: str = field(init=False)

    def __post_init__(self) -> None:
        invalid = False
        try:
            if not all(isinstance(item, str) and len(item) <= 128 and item.strip() and len(item.encode("utf-8")) <= 128
                       for item in (self.schema_id, self.schema_version)):
                raise ValueError("invalid schema identity")
            if self.schema_profile != OUTPUT_SCHEMA_PROFILE:
                raise ValueError("unsupported schema profile")
            if not isinstance(self.limits, StructuredOutputLimits):
                raise ValueError("invalid limits")
            schema = _snapshot(self.schema, depth=self.limits.schema_depth,
                               nodes=self.limits.schema_nodes, byte_limit=self.limits.schema_bytes)
            if not isinstance(schema, dict) or schema.get("type") != "object":
                raise ValueError("root must be object")
            _check_schema_nodes(schema)
            if schema_contract_violations(schema, path="$"):
                raise ValueError("invalid schema")
        except (ValueError, TypeError, OverflowError, RecursionError):
            invalid = True
        if invalid:
            raise _failure("structured_output_schema_invalid", "schema")
        if self.mode not in ("local", "native_required"):
            raise _failure("structured_output_mode_unsupported", "mode")
        object.__setattr__(self, "schema", freeze_json_mapping(schema))
        object.__setattr__(self, "schema_digest", _digest(schema))
        object.__setattr__(self, "contract_digest", _digest({
            "schemaId": self.schema_id, "schemaVersion": self.schema_version,
            "schemaProfile": self.schema_profile, "schemaDigest": self.schema_digest,
            "mode": self.mode, "limits": list(getattr(self.limits, key) for key in self.limits.__dataclass_fields__),
        }))

    def identity(self) -> Mapping[str, Any]:
        return freeze_json_mapping({
            "schemaId": self.schema_id, "schemaVersion": self.schema_version,
            "schemaProfile": self.schema_profile, "schemaDigest": self.schema_digest,
            "contractDigest": self.contract_digest, "mode": self.mode,
        })

    def schema_json(self) -> str:
        from purra.json_values import thaw_json_mapping
        return json.dumps(thaw_json_mapping(self.schema), ensure_ascii=False, separators=(",", ":"))

    def instruction(self) -> str:
        header = "Return exactly one complete JSON object. Do not use tools, markdown fences, or extra text."
        return header + (" JSON Schema: " + self.schema_json() if self.mode == "local" else " Follow the native JSON Schema output format.")

    def parse(self, text: str) -> Mapping[str, Any]:
        """Validate a complete JSON document and return a detached frozen object."""
        invalid = False
        try:
            if not isinstance(text, str) or len(text) > self.limits.output_bytes:
                raise ValueError("invalid document")
            if len(text.encode("utf-8")) > self.limits.output_bytes:
                raise ValueError("document too large")
            # Guard nesting before the native parser allocates a recursive tree.
            level = 0
            quoted = escaped = False
            for char in text:
                if quoted:
                    if escaped:
                        escaped = False
                    elif char == "\\":
                        escaped = True
                    elif char == '"':
                        quoted = False
                elif char == '"':
                    quoted = True
                elif char in "[{":
                    level += 1
                    if level > self.limits.output_depth + 1:
                        raise ValueError("document too deep")
                elif char in "]}":
                    level -= 1

            def pairs(items):
                result = {}
                for key, value in items:
                    if key in result:
                        raise ValueError("duplicate key")
                    result[key] = value
                return result

            def constant(_):
                raise ValueError("non-JSON number")

            value = json.loads(text, object_pairs_hook=pairs, parse_constant=constant)
            value = _snapshot(value, depth=self.limits.output_depth,
                              nodes=self.limits.output_nodes, byte_limit=self.limits.output_bytes)
        except (ValueError, TypeError, OverflowError, RecursionError):
            invalid = True
        if invalid:
            raise _failure("structured_output_invalid_json", "document")
        return self.validate_value(value)

    def validate_value(self, value: Any) -> Mapping[str, Any]:
        """Validate an already-decoded object with the same bounded profile."""
        try:
            value = _snapshot(value, depth=self.limits.output_depth,
                              nodes=self.limits.output_nodes, byte_limit=self.limits.output_bytes)
        except (ValueError, TypeError, OverflowError, RecursionError):
            raise _failure("structured_output_invalid_json", "value") from None
        try:
            violation = schema_violation(value, self.schema, path="$", work=[self.limits.validation_steps])
        except ValueError:
            raise _failure("structured_output_validation_limit_exceeded", "validation_steps") from None
        if violation is not None:
            details = violation[1]
            # Paths derived from undeclared candidate keys never enter diagnostics.
            path = details["path"] if details["keyword"] != "additionalProperties" else "$"
            raise _failure("structured_output_schema_mismatch", details["keyword"], path)
        return freeze_json_mapping(value)


def json_identity_digest(value: Any) -> str:
    """Hash bounded interoperable JSON using purra.json-identity/v1."""
    limits = StructuredOutputLimits()
    normalized = _snapshot(value, depth=limits.output_depth, nodes=limits.output_nodes,
                           byte_limit=limits.output_bytes)
    return _digest(normalized)


__all__ = ["OUTPUT_SCHEMA_PROFILE", "JSON_IDENTITY_PROFILE", "StructuredOutputContract",
           "StructuredOutputLimits", "StructuredOutputError", "json_identity_digest"]
